### ISSUE TITLE
MongoDB not starting

### ISSUE THREAD

**Issue by aadi-khan** [cytopia/devilbox#160]
Hi,

I am using the windows 10 pro and i have recently installed the docker the latest version.

I wanted to use devilbox services everything is working fine but the mongodb and postgre database.

I am getting the below mention error while looking into the mongo logs.
```
E:\Docker\devilbox (master)
λ docker-compose logs mongo
Attaching to devilbox_mongo_1
mongo_1  | 2017-12-05T15:07:52.781+0000 I CONTROL  [initandlisten] MongoDB starting : pid=1 port=27017 dbpath=/data/db 64-bit host=4b8802b75a60
mongo_1  | 2017-12-05T15:07:52.781+0000 I CONTROL  [initandlisten] db version v3.4.10
mongo_1  | 2017-12-05T15:07:52.781+0000 I CONTROL  [initandlisten] git version: 078f28920cb24de0dd479b5ea6c66c644f6326e9
mongo_1  | 2017-12-05T15:07:52.781+0000 I CONTROL  [initandlisten] OpenSSL version: OpenSSL 1.0.1t  3 May 2016
mongo_1  | 2017-12-05T15:07:52.781+0000 I CONTROL  [initandlisten] allocator: tcmalloc
mongo_1  | 2017-12-05T15:07:52.781+0000 I CONTROL  [initandlisten] modules: none
mongo_1  | 2017-12-05T15:07:52.781+0000 I CONTROL  [initandlisten] build environment:
mongo_1  | 2017-12-05T15:07:52.781+0000 I CONTROL  [initandlisten]     distmod: debian81
mongo_1  | 2017-12-05T15:07:52.781+0000 I CONTROL  [initandlisten]     distarch: x86_64
mongo_1  | 2017-12-05T15:07:52.781+0000 I CONTROL  [initandlisten]     target_arch: x86_64
mongo_1  | 2017-12-05T15:07:52.781+0000 I CONTROL  [initandlisten] options: {}
mongo_1  | 2017-12-05T15:07:52.788+0000 I STORAGE  [initandlisten] wiredtiger_open config: create,cache_size=478M,session_max=20000,eviction=(threads_min=4,threads_max=4),config_base=false,statistics=(fast),log=(enabled=true,archive=true,path=journal,compressor=snappy),file_manager=(close_idle_time=100000),checkpoint=(wait=60,log_size=2GB),statistics_log=(wait=0),
mongo_1  | 2017-12-05T15:07:52.954+0000 E STORAGE  [initandlisten] WiredTiger error (17) [1512486472:954333][1:0x7f1585c39d40], connection: /data/db/WiredTiger.wt: handle-open: open: File exists
mongo_1  | 2017-12-05T15:07:52.956+0000 I STORAGE  [initandlisten] WiredTiger message unexpected file WiredTiger.wt found, renamed to WiredTiger.wt.11
mongo_1  | 2017-12-05T15:07:52.958+0000 E STORAGE  [initandlisten] WiredTiger error (1) [1512486472:958291][1:0x7f1585c39d40], connection: /data/db/WiredTiger.wt: handle-open: open: Operation not permitted
mongo_1  | 2017-12-05T15:07:52.960+0000 I -        [initandlisten] Assertion: 28595:1: Operation not permitted src/mongo/db/storage/wiredtiger/wiredtiger_kv_engine.cpp 276
mongo_1  | 2017-12-05T15:07:52.962+0000 I STORAGE  [initandlisten] exception in initAndListen: 28595 1: Operation not permitted, terminating
mongo_1  | 2017-12-05T15:07:52.963+0000 I NETWORK  [initandlisten] shutdown: going to close listening sockets...
mongo_1  | 2017-12-05T15:07:52.963+0000 I NETWORK  [initandlisten] removing socket file: /tmp/mongodb-27017.sock
mongo_1  | 2017-12-05T15:07:52.963+0000 I NETWORK  [initandlisten] shutdown: going to flush diaglog...
mongo_1  | 2017-12-05T15:07:52.963+0000 I CONTROL  [initandlisten] now exiting
mongo_1  | 2017-12-05T15:07:52.963+0000 I CONTROL  [initandlisten] shutting down with code:100
```
I dont understand this error and so i decided to ask you for help. Please guide me if there anything i need to change.

Thank you very much

**Comment by cytopia** [comment]
@aadi-khan can you please post your `.env` file as well as docker and docker-compose versions

**Comment by aadi-khan** [comment]
Sure,

Here is the files .env and docker-compose in the zip folder.

[devilbox.zip](https://github.com/cytopia/devilbox/files/1531765/devilbox.zip)

**Comment by aadi-khan** [comment]
![screenshot-localhost-2017-12-05-20-44-35-954](https://user-images.githubusercontent.com/7964591/33615947-41c65636-d9fd-11e7-97ee-9e09a6315518.png)

This is the screenshot of the devilbox page, it shows 100% health but two services have not yet started.

**Comment by cytopia** [comment]
Please provide Docker and Docker Compose versions and also your command you used to start up the devilbox

**Comment by aadi-khan** [comment]
Docker version 17.09.0-ce, build afdb6d4

docker-compose version 1.16.1, build 6d1ac219

I have download the devilbox repository and after making the .env file and i have simply run docker-compose up -d

**Comment by cytopia** [comment]
Are you using Docker for Windows or Docker Toolbox?

**Comment by aadi-khan** [comment]
I am using docker for windows not the toolbox

**Comment by cytopia** [comment]
Can you shutdown the devilbox and try a few different MongoDB versions and check if it has the same problem

Properly shutdown before starting up a different version:
```shell
docker-compose down
docker-compose rm -f
```

**Comment by aadi-khan** [comment]
Alright, I am now shutting down my services and restart again with the different mongodb version. I am currently using MONGO_SERVER=3.4 and now i am going to use the MONGO_SERVER=3.5 and than i will let you know if this happen again.

Thank you very much for your time.

**Comment by cytopia** [comment]
Just FYI: If you do not need MongoDB, you could also just startup exactly what you need.

**Example (Only starts up httpd, bind, php and mysql):**
```
docker-compose up httpd bind php mysql
```

**Comment by aadi-khan** [comment]
Yes I have already setup wordpress and its working fine but i have few more projects that i need to setup which is angular and magento. I wanna bind the separate angular container with the mongodb. Is it possible ?

**Comment by aadi-khan** [comment]
Nope its not working either on mongodb 3.5, I think i should have to setup the another MEAN stack container stack.

Thanks again bro

**Comment by cytopia** [comment]
@aadi-khan could you try one more thing to check if the MongoDB actually works on Windows:

Without mounts
```
docker run -it -p 27017:27017 mongo:3.4
```

With mounts
```
docker run -it -p 27017:27017 -v /tmp/mongo:/data/db  mongo:3.4
```

**Comment by aadi-khan** [comment]
This is the message on without mounts

**Comment by aadi-khan** [comment]
λ docker run -it -p 27017:27017 mongo:3.5
2017-12-06T07:02:03.279+0000 I CONTROL  [initandlisten] MongoDB starting : pid=1 port=27017 dbpath=/data/db 64-bit host=dc0a689a9e68
2017-12-06T07:02:03.279+0000 I CONTROL  [initandlisten] db version v3.5.13
2017-12-06T07:02:03.279+0000 I CONTROL  [initandlisten] git version: 52bbaa007cd84631d6da811d9a05b59f2dfad4f3
2017-12-06T07:02:03.280+0000 I CONTROL  [initandlisten] OpenSSL version: OpenSSL 1.0.1t  3 May 2016
2017-12-06T07:02:03.280+0000 I CONTROL  [initandlisten] allocator: tcmalloc
2017-12-06T07:02:03.280+0000 I CONTROL  [initandlisten] modules: none
2017-12-06T07:02:03.280+0000 I CONTROL  [initandlisten] build environment:
2017-12-06T07:02:03.281+0000 I CONTROL  [initandlisten]     distmod: debian81
2017-12-06T07:02:03.281+0000 I CONTROL  [initandlisten]     distarch: x86_64
2017-12-06T07:02:03.281+0000 I CONTROL  [initandlisten]     target_arch: x86_64
2017-12-06T07:02:03.281+0000 I CONTROL  [initandlisten] options: { net: { bindIpAll: true } }
2017-12-06T07:02:03.281+0000 I STORAGE  [initandlisten]
2017-12-06T07:02:03.281+0000 I STORAGE  [initandlisten] ** WARNING: Using the XFS filesystem is strongly recommended with the WiredTiger storage engine
2017-12-06T07:02:03.281+0000 I STORAGE  [initandlisten] **          See http://dochub.mongodb.org/core/prodnotes-filesystem
2017-12-06T07:02:03.282+0000 I STORAGE  [initandlisten] wiredtiger_open config: create,cache_size=478M,session_max=20000,eviction=(threads_min=4,threads_max=4),config_base=false,statistics=(fast),log=(enabled=true,archive=true,path=journal,compressor=snappy),file_manager=(close_idle_time=100000),statistics_log=(wait=0),verbose=(recovery_progress),
2017-12-06T07:02:04.713+0000 I CONTROL  [initandlisten]
2017-12-06T07:02:04.713+0000 I CONTROL  [initandlisten] ** NOTE: This is a development version (3.5.13) of MongoDB.
2017-12-06T07:02:04.713+0000 I CONTROL  [initandlisten] **       Not recommended for production.
2017-12-06T07:02:04.713+0000 I CONTROL  [initandlisten]
2017-12-06T07:02:04.714+0000 I CONTROL  [initandlisten] ** WARNING: Access control is not enabled for the database.
2017-12-06T07:02:04.714+0000 I CONTROL  [initandlisten] **          Read and write access to data and configuration is unrestricted.
2017-12-06T07:02:04.714+0000 I CONTROL  [initandlisten]
2017-12-06T07:02:05.549+0000 I FTDC     [initandlisten] Initializing full-time diagnostic data capture with directory '/data/db/diagnostic.data'
2017-12-06T07:02:05.923+0000 I COMMAND  [initandlisten] setting featureCompatibilityVersion to 3.6
2017-12-06T07:02:05.924+0000 I NETWORK  [initandlisten] waiting for connections on port 27017

**Comment by aadi-khan** [comment]
And this is the message appears with mounts , it say the port is already allocated 

E:\Docker\devilbox (master)
λ docker run -it -p 27017:27017 -v /tmp/mongo:/data/db  mongo:3.5
docker: Error response from daemon: driver failed programming external connectivity on endpoint vigilant_wilson (b7eb68a75788bb47ae0abd4ca4993671842c5aae2286dade3ad85df12b18a9b2): Bind for 0.0.0.0:27017 failed: port is already allocated.
ERRO[0001] error waiting for container: context canceled

E:\Docker\devilbox (master)
λ docker run -it -p 27017:27017 -v /tmp/mongo:/data/db  mongo:3.4
docker: Error response from daemon: driver failed programming external connectivity on endpoint reverent_mirzakhani (b2e9e500d93a14f6d780664ce5440f13fc24b0d3fff21a46216cde6963e737ed): Bind for 0.0.0.0:27017 failed: port is already allocated.

**Comment by aadi-khan** [comment]
Its just keep waiting on both situation with mounts and without mounts.

**Comment by cytopia** [comment]
The first one (without mounts) seems fine as far as I can tell. It looks like initialization was successful and it awaits connection.

The second one cannot start up, as the port is already in use (maybe the first run wasn't shutdown properly and is still running?). Can you check (before trying the mount version) with `docker ps` if you already have a mongo db running, or leave out the port completely:
```
docker run -it  -v /tmp/mongo:/data/db mongo:3.4
```

**Comment by aadi-khan** [comment]
Now i have stopped all the containers and restart the docker-compose up -d.

λ docker-compose up -d
Creating network "devilbox_app_net" with driver "bridge"
Creating devilbox_bind_1 ...
Creating devilbox_bind_1 ... done
Creating devilbox_php_1 ...
Creating devilbox_php_1 ... done
Creating devilbox_httpd_1 ...
Creating devilbox_httpd_1 ... done
Creating devilbox_mysql_1 ...
Creating devilbox_pgsql_1 ...
Creating devilbox_redis_1 ...
Creating devilbox_mongo_1 ...
Creating devilbox_memcd_1 ...
Creating devilbox_mysql_1
Creating devilbox_pgsql_1
Creating devilbox_redis_1
Creating devilbox_mongo_1
Creating devilbox_memcd_1 ... done

Everything is up now but still the mongodb does not start on the localhost page

**Comment by aadi-khan** [comment]
it says the health is 100% but the 2 services havent yet started

**Comment by aadi-khan** [comment]
nor the Mongo container appears on docker ps

**Comment by cytopia** [comment]
Can you give a standalone MongoDB container with mounts and withouts ports a try:
```
docker run -it  -v /tmp/mongo:/data/db mongo:3.4
```

**Comment by cytopia** [comment]
Also make sure  you have the latest MongoDB images:
```
docker pull mongo:3.4
docker pull mongo:3.5
```

**Comment by aadi-khan** [comment]
λ docker images
REPOSITORY              TAG                 IMAGE ID            CREATED             SIZE
redis                   4.0                 e5f9a9e182f4        4 days ago          107MB
postgres                9.6                 d3ac03f9698d        3 weeks ago         268MB
mongo                   3.5                 a2ab5e364821        4 weeks ago         371MB
mongo                   3.4                 d22888af0ce0        4 weeks ago         361MB
memcached               1.5.2               5dceb668fb20        4 weeks ago         58.6MB
devilbox/nginx-stable   0.12                da8e151ed02a        2 months ago        148MB
cytopia/php-fpm-7.1     latest              927ad858fb6a        2 months ago        1.1GB
cytopia/mysql-8.0       latest              32897f5dddb5        3 months ago        2GB
cytopia/bind            latest              d0479e969008        3 months ago        198MB
d4w/nsenter             latest              9e4f13a0901e        14 months ago       83.8kB

**Comment by aadi-khan** [comment]
![untitled](https://user-images.githubusercontent.com/7964591/33650643-acf9190c-da84-11e7-8bb2-b538be7b00f2.png)

**Comment by cytopia** [comment]
Seems to work fine standalone.

Can you stop all containers:
```
docker ps -q | xargs docker stop
```
Shutdown the devilbox
```
docker-compose stop; docker-compose rm -f
```

Comment the following line in `docker-compose.yml` (around line 413), so that MongoDB does not mount anything
```
      #- ${HOST_PATH_MONGO_DATADIR}/${MONGO_SERVER}:/data/db
```

And then start up just a few containers:
```
docker-compose up httpd bind php mongo
```

**Comment by aadi-khan** [comment]
λ docker-compose up httpd bind php mongo
ERROR: The Compose file '.\docker-compose.yml' is invalid because:
services.mongo.volumes contains an invalid type, it should be an array

**Comment by cytopia** [comment]
Also comment the `volumes` line in the `mongo:` section:
```
    #volumes:
```

**Comment by aadi-khan** [comment]
![untitled](https://user-images.githubusercontent.com/7964591/33651390-801d3c94-da87-11e7-905e-0c7df478d52d.png)

seems like its working but on localhost it get 503 error gateway timeout

**Comment by aadi-khan** [comment]
![untitled1](https://user-images.githubusercontent.com/7964591/33651510-eb2aff26-da87-11e7-96ce-fc39e90b91fc.png)

**Comment by cytopia** [comment]
Looks like it is receiving connections now. Are you still getting the `504` on `http://localhost`? If so, can you paste the web server error  logs here.

**Comment by aadi-khan** [comment]
![untitled2](https://user-images.githubusercontent.com/7964591/33651723-ae84fa12-da88-11e7-985f-6ea51c5403bf.png)

Yes still get the 504 gateway timeout error

**Comment by cytopia** [comment]
Can you provide the logfiles from `log/nginx-<your-version>/*`

**Comment by aadi-khan** [comment]
[defaultlocalhost-access.log](https://github.com/cytopia/devilbox/files/1534384/defaultlocalhost-access.log)
[defaultlocalhost-error.log](https://github.com/cytopia/devilbox/files/1534385/defaultlocalhost-error.log)

**Comment by aadi-khan** [comment]
did you get the errors, what should i do now what would you suggest ?

Thanks

**Comment by cytopia** [comment]
Looks like it cannot find `index.php`. Did you remove any other volumes from `docker-compose.yml`?

**Comment by aadi-khan** [comment]
hey its working fine now, all the services are running now i comment the volume part of postgre and it work as well as mongo.

Thank you so much for your time and support you are amazing.

**Comment by cytopia** [comment]
Ok, however this is still not the optimal solution as data inside MongoDB/PostgreSQL might not persist across devilbox restarts. At least we know that it does work and seem to have some problem with paths on Windows mounts.

If I come up with a new idea I will let you know.

**Comment by aadi-khan** [comment]
Alright, wish you all the best for new idea's.

Just one last thing if you dont mind "what is your name sir ?"

**Comment by cytopia** [comment]
@aadi-khan just call me by my github nickname here :-)

Btw, could you paste all the changes here you now have applied to `docker-compose.yml`.

```
git diff origin/master -- docker-compose.yml
```

**Comment by aadi-khan** [comment]
I have just comment out the MongoDB volume section 

# ------------------------------------------------------------
  # MongoDB
  # ------------------------------------------------------------
  mongo:
    image: mongo:${MONGO_SERVER:-latest}

    ports:
      # [local-machine:]local-port:docker-port
      - "${LOCAL_LISTEN_ADDR}${HOST_PORT_MONGO}:27017"

    networks:
      app_net:
        ipv4_address: 172.16.238.16

    # volumes:
      # ---- Format: ----
      # HOST-DIRECTORY : DOCKER-DIRECTORY

      # Mount MongoDB Data directory
      # - ${HOST_PATH_MONGO_DATADIR}/${MONGO_SERVER}:/data/db

    depends_on:
      - bind
      - php
      - httpd

Same i did with Postgre

# ------------------------------------------------------------
  # PostgreSQL
  # ------------------------------------------------------------
  pgsql:
    image: postgres:${PGSQL_SERVER:-9.6}

    environment:

      - POSTGRES_USER=${PGSQL_ROOT_USER}
      - POSTGRES_PASSWORD=${PGSQL_ROOT_PASSWORD}
      - PGDATA=/var/lib/postgresql/data/pgdata

    ports:
      # [local-machine:]local-port:docker-port
      - "${LOCAL_LISTEN_ADDR}${HOST_PORT_PGSQL}:5432"

    networks:
      app_net:
        ipv4_address: 172.16.238.13

    # volumes:
      # ---- Format: ----
      # HOST-DIRECTORY : DOCKER-DIRECTORY

      # Mount logs
      # - ${DEVILBOX_PATH}/log/pgsql-${PGSQL_SERVER}:/var/log/postgresql

      # Mount PostgreSQL Data directory
      # - ${HOST_PATH_PGSQL_DATADIR}/${PGSQL_SERVER}:/var/lib/postgresql/data/pgdata

    depends_on:
      - bind
      - php
      - httpd

**Comment by aadi-khan** [comment]
I have encountered  another issue that is related to php, first thing is i am not able to access the shell.sh and the second is laravel installer throws the warnings and errors while i was executing the command 
> $ docker-compose exec --user devilbox php laravel new laravel

In short i wasnt able to use the php tools.

**Comment by cytopia** [comment]
Just go with `./shell.sh` and then post the errors here you encounter with the `laravel` cli-tool.

**Comment by aadi-khan** [comment]
![untitled](https://user-images.githubusercontent.com/7964591/33715358-08397182-db74-11e7-9372-73e8d071117c.png)

when i run _shell.sh_ this window pops up and disappear and nothing works after it.

**Comment by aadi-khan** [comment]
also is it possible to install Mysql and intl php extensions in my php module ?

**Comment by aadi-khan** [comment]
![untitled](https://user-images.githubusercontent.com/7964591/33715521-ae573b1c-db74-11e7-96c5-66630ba4b208.png)

See this is what happen when i run the exec command

**Comment by aadi-khan** [comment]
![untitled](https://user-images.githubusercontent.com/7964591/33715649-205c518e-db75-11e7-8e42-46b022ab2619.png)

This is what appears when i run the exec command in htdocs folder

**Comment by aadi-khan** [comment]
I would appreciate if you can confirm that the php extensions can be install in the current module or no.

Thank you

**Comment by cytopia** [comment]
MySQL and intl extensions are installed already. However it looks like you have a seriously broken configuration. All those warnings/errors shown in your screenshot above should never appear.

Can you make a fresh clone of the devilbox (even use a different directory name for the clone) and see if all those warnings go away.

**Comment by cytopia** [comment]
Sorry my bad. Just saw that you are executing `docker-compose` in the `htdocs/` directory. You can only ever use `docker-compose` command within the root git directory. It requires the `docker-compose.yml` to be in the same directory.

For what you want to do, simply use `shell.sh` to enter the Docker container and just execute your commands normally:
```
$ ./shell.sh
```
https://github.com/cytopia/devilbox/blob/master/docs/Usage.md#3-work-inside-the-php-container

**Comment by destillat** [comment]
@aadi-khan you probably forgot to mount your E drive
https://docs.docker.com/docker-for-windows/#shared-drives

**Comment by cytopia** [comment]
@aadi-khan could you please report back if the suggestion form @destillat works for you.

**Comment by cytopia** [comment]
I will close this due to inactivity. In case you still have problems, please checkout this branch if it is still open at the time of reading: https://github.com/cytopia/devilbox/pull/203

If this is already merged to master, simply update your local git directory
